### PR TITLE
Fix SURF OCL

### PR DESCRIPTION
**Pull request by acyen**
Enable SURF OCL. Fix incorrect calling sequence for two kernels.
Fix compiler warnings in surf.cl.
Enable SURF OCL test. Test runs with no failures.

### CODE REVIEW

**Comment by alalek** [comment]
Please take a look on existed OpenCV tests. They are failed on SURF checks.

**Comment by acyen** [comment]
What is the right way to fix this test - DescriptorExtractor_SURF? The test is comparing the CPU descriptor calculation with OCL.  I fixed the OCL code to match CUDA, but the CPU code is different from both and gives a different result. Choices are:
1. Create a separate test for OCL and do not compare CPU with OCL.
2. Change CPU code to match OCL.
3. Change OCL code to match CPU.

**Comment by catree** [comment]
My comments, as just an OpenCV user. The best would be in my opinion:

- for the descriptor part: have a numerically accurate code + without decrease in term of precision / recall

- match the CPU and GPU code if possible

Looks like it is not so simple (only bullet 3 should concern this PR):

- detection part of OCL SURF seems to be of less quality: [OpenCL SURF matching quality decreased #4685](https://github.com/opencv/opencv/issues/4685)

- Cuda SURF seems to suffer of the same issue (at least for the detection part): [SURF_CUDA produces different number of keypoints on the same image #584](https://github.com/opencv/opencv_contrib/issues/584)

- one user reports NaN values with Cuda SURF in some cases: [SURF_CUDA produces nan values](http://answers.opencv.org/question/130616/surf_cuda-produces-nan-values/)

- even if the OCL / Cuda descriptor part matches the CPU descriptor code, the Cuda detection part seems to be [not deterministic](https://github.com/opencv/opencv_contrib/issues/584). It is not necessary an issue though, as the GPU architecture is of course different of a CPU architecture.

To recap, as a user the best would be option 2 if there is no decrease in term of performance and quality (precision / recall) for CPU SURF.

PS: I found an old [SURF benchmark](http://boofcv.org/index.php?title=Performance:SURF).

**Comment by acyen** [comment]
Changing CPU SURF would break tests unless the test data were changed at the same time.  Is there a procedure to do this?

**Comment by catree** [comment]
I think that you will have to submit a PR with the new test data [here](https://github.com/opencv/opencv_extra).

**Comment by acyen** [comment]
The testdata PR has to be coordinated with this one, otherwise tests will break.  How does that work?

**Comment by alalek** [comment]
You need to use the same branch name for patches in both forks (opencv_contrib/opencv_extra).
But this doesn't work for "master" source branch(see above: "from acyen:master"). You need to recreate PR with another branch name (for example, surf_ocl).
